massa-bootstrap/src/listener.rs

```diff
@@ -6,69 +6,79 @@ use mio::{Events, Interest, Poll, Token, Waker};
 use tracing::{error, info};
 
 use crate::error::BootstrapError;
+use crate::establisher::BSListener;
 
 const NEW_CONNECTION: Token = Token(0);
 const STOP_LISTENER: Token = Token(10);
 
-pub(crate) struct BootstrapTcpListener();
+/// TODO: this should be crate-private. currently needed for models testing
+pub struct BootstrapTcpListener {
+    poll: Poll,
+    events: Events,
+    server: TcpListener,
+    // HACK : create variable to move ownership of mio_server to the thread
+    // if mio_server is not moved, poll does not receive any event from listener
+    _mio_server: MioTcpListener,
+}
 
 pub struct BootstrapListenerStopHandle(Waker);
 
+pub enum AcceptEvent {
+    NewConnection((TcpStream, SocketAddr)),
+    Stop,
+}
 impl BootstrapTcpListener {
     /// Start a new bootstrap listener on the given address.
     ///
     /// * `addr` - the address to listen on
     /// * `connection_tx` - the channel to send new connections to
-    pub fn start(
-        addr: SocketAddr,
-        connection_tx: crossbeam::channel::Sender<(TcpStream, SocketAddr)>,
-    ) -> Result<BootstrapListenerStopHandle, BootstrapError> {
+    pub fn new(addr: SocketAddr) -> Result<(BootstrapListenerStopHandle, Self), BootstrapError> {
         info!("Starting bootstrap listener on {}", &addr);
         let server = TcpListener::bind(addr)?;
         let mut mio_server =
             MioTcpListener::from_std(server.try_clone().expect("Unable to clone server socket"));
 
-        let mut poll = Poll::new()?;
+        let poll = Poll::new()?;
 
         // wake up the poll when we want to stop the listener
-        let waker = Waker::new(poll.registry(), STOP_LISTENER)?;
+        let waker = BootstrapListenerStopHandle(Waker::new(poll.registry(), STOP_LISTENER)?);
 
         poll.registry()
             .register(&mut mio_server, NEW_CONNECTION, Interest::READABLE)?;
 
         // TODO use config for capacity ?
-        let mut events = Events::with_capacity(32);
-
-        // spawn a thread to handle events
-        std::thread::Builder::new()
-            .name("bs_listener".to_string())
-            .spawn(move || {
-                // HACK : create variable to move ownership of mio_server to the thread
-                // if mio_server is not moved, poll does not receive any event from listener
-                let _temp = mio_server;
+        let events = Events::with_capacity(32);
+        Ok((
+            waker,
+            BootstrapTcpListener {
+                poll,
+                server,
+                events,
+                _mio_server: mio_server,
+            },
+        ))
+    }
+}
 
-                loop {
-                    poll.poll(&mut events, None).unwrap();
+impl BSListener for BootstrapTcpListener {
+    fn accept(&mut self) -> Result<AcceptEvent, BootstrapError> {
```

Comment:
i'm not a big fan of no thread the poll event, STOP_LISTENER is called only when we call `accept()` ?

Reply:
the stop-listener and the connection listener both get polled for. You can see the stop-waker being registered here: https://github.com/massalabs/massa/pull/3819/files#diff-75fc005ad20f9d19fed3694ee99b9d4e31d177cc1462e438607552996b8b823bR41-R61

The idea being, is that the main-loop has direct access to the `BsListener`. in the caso of using a `BootstrapTcpListener` in this Pr, you have access to:

- a `Poll`, that has a waker registered to it
- a TcpListener that is registered to the same poll

This means that the TcpListener can be written in such a way that it blocks until either the waker is signaled, or the listener can accept a connection, and is able to differentiate between the two.

Reply:
In short: It's as if we could just use a a std::net::TcpListener like so:

```rust
let stoppable_listener = std::net::TcpListener::bind(...).with_stopper(&stopper_thing);
let server_thread =  std::thread::spawn(|| loop {

    let conn = match stopable_listener.stopable_accept() {
        Ok(Event::Accept(conn)) => conn,
        Ok(Event::stop => {
            info!("system-wide stopper invoked. exiting cleanly");
            return;
        },
        Err(e) => panic!("tcp listener errored out"),
    };
    // do thing with connection
});
// wait untill the stopper thing does its thing so the thread returns out
server_thread.join().unwrap();
info!("server succesfully shutdown gracefully after system-wide stop-signal invoked")
```

Reply:
Ok we can go with that; 

Maybe stopable_accept can be rename to accept() wich return Err(...) when the stop is called. This allow to remove the Event and has a cleaner result.

Reply:
the code i replied with is just an illustrative example. `stopable_accept` is a hypothetical method that would mirror the behavior as our current accept. No difference, other than the fact that the result can be "stop" or "here's the next connection" instead of _just_ "here's the next connection" (assuming no error) 

If something is an Err(...), that implies an incorrect, but handleable, event has occured. Trying to open a file that doesn't exist, for example. 

In this case, it's perfectly expected to get a Stop.